### PR TITLE
Model Training Improvements

### PR DESCRIPTION
**Pull request by aryamansriram**
This branch contains added balanced batching support and Early stopping to the binary relevance model script

### CODE REVIEW

**Comment by aryamansriram** [comment]
[Model training improvements](https://trello.com/c/WkfQhWrE/32-model-training-improvements)

File: models/multilabel_model.py [hunk 1]
```diff
@@ -322,11 +318,23 @@ def create_keras_model():
...
 
+callback = EarlyStopping(
+    monitor="val_loss",
+    verbose=1,
+    mode="auto"
+)
 
-MODEL_TRAINING = MODEL.fit(CLF2_TRAIN, CLF2_TRAIN_TARGET,
+training_generator = make_generator(
+    CLF2_TRAIN, CLF2_TRAIN_TARGET.values, batch_size=batch_size, categorical=True)
```
Comment (nandan-sridhar) [inline models/multilabel_model.py:328]:
```suggestion
training_generator = make_generator(
    CLF2_TRAIN, CLF2_TRAIN_TARGET.values, batch_size=batch_size, categorical=True, seed=42)
```

File: models/binary_relevance_model.py [hunk 2]
```diff
@@ -367,8 +372,18 @@ def create_keras_model():
...
 #############################################################################
+
+callback = EarlyStopping(
+    monitor="val_accuracy",
+    verbose=1,
+    mode="max"
+)
+
+training_generator = make_generator(
+    CLF2_TRAIN, to_categorical(CLF2_TRAIN_TARGET), batch_size=CONFIG_DATA["train"]["batchSize"], categorical=False)
```
Comment (nandan-sridhar) [inline models/binary_relevance_model.py:385]:
```suggestion
training_generator = make_generator(
    CLF2_TRAIN, to_categorical(CLF2_TRAIN_TARGET), batch_size=CONFIG_DATA["train"]["batchSize"], categorical=False, seed=42)
```

File: models/binary_relevance_model.py [hunk 3]
```diff
@@ -377,15 +392,29 @@ def create_keras_model():
...
                            class_weight={0: CLASS_WEIGHT_0, 1: CLASS_WEIGHT_1},
                            verbose=1,
-                           validation_data=(CLF2_TEST, CLF2_TEST_TARGET))
+                           callbacks=[callback],
+                           validation_data=(CLF2_TEST, CLF2_TEST_TARGET))'''
+
+steps_per_epoch = len(CLF2_TRAIN) // CONFIG_DATA["train"]["batchSize"]
+MODEL_TRAINING = MODEL.fit(training_generator, shuffle=True, epochs=100, steps_per_epoch=steps_per_epoch,
+                           callbacks=[callback],
+                           validation_data=(CLF2_TEST, CLF2_TEST_TARGET.reshape(-1)), verbose=1)
```
Comment (nandan-sridhar) [inline models/binary_relevance_model.py:409]:
```suggestion
MODEL_TRAINING = MODEL.fit(training_generator, shuffle=True,
                           epochs=CONFIG_DATA["train"]["epochs"],
                           steps_per_epoch=steps_per_epoch,
                           callbacks=[callback],
                           validation_data=(CLF2_TEST, CLF2_TEST_TARGET.reshape(-1)), verbose=1)
```

File: models/binary_relevance_model.py [hunk 4]
```diff
@@ -377,15 +392,29 @@ def create_keras_model():
     loaded_model_json = json_file.read()
     json_file.close()
     MODEL = model_from_json(loaded_model_json)
-    MODEL.compile(loss='binary_crossentropy', optimizer=Adam(lr=1e-4, epsilon=1e-8),
+    MODEL.compile(loss='binary_crossentropy', optimizer=Adam(lr=1e-5, epsilon=1e-8),
```
Comment (nandan-sridhar) [inline models/binary_relevance_model.py:395]: Please also make similar change to learning rate in multilabel_model.py. Also for create_keras_model functions in both files, lr should be 1e-5. So total 3 places

File: models/binary_relevance_model.py [hunk 5]
```diff
@@ -377,15 +392,29 @@ def create_keras_model():
...
 
-MODEL_TRAINING = MODEL.fit(CLF2_TRAIN, CLF2_TRAIN_TARGET,
+'''MODEL_TRAINING = MODEL.fit(CLF2_TRAIN, CLF2_TRAIN_TARGET,
                            epochs=CONFIG_DATA["train"]["epochs"],
                            batch_size=CONFIG_DATA["train"]["batchSize"],
                            class_weight={0: CLASS_WEIGHT_0, 1: CLASS_WEIGHT_1},
                            verbose=1,
-                           validation_data=(CLF2_TEST, CLF2_TEST_TARGET))
+                           callbacks=[callback],
+                           validation_data=(CLF2_TEST, CLF2_TEST_TARGET))'''
```
[the remark this answers is not in the document]
Comment (aryamansriram) [reply models/binary_relevance_model.py:403]: Done

**Pull request merged**